### DISCUSSION TITLE
指で画像を動かす

### DISCUSSION THREAD

**Discussion by ShunHigu** [lit-iphone-course/lit-iphone-course#84] [Question-original] [answered]
１　配列で宣言した複数の画像を指で動かすコードを書きたいです。

２　サイトにあったようにStoneImageとSekkiimageをSozaiViewとして読み込ませる方法を取りました。
Fatal error: Index out of rangeのエラーが出ないようにするためにif文を用いているのですが、エラーが発生してしまいます。エラーが出ないようにするためには、どのように変更すれば良いでしょうか？

3　複数画像を指で動かすサイト
https://swiswiswift.com/2022-06-04/

４
override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?){
        
        if stonenum>=1{
            for i in 1...stonenum{
                StoneImage[i]=SozaiView()
            }
        }
        
        if sekkinum>=1{
            for i in 1...stonenum{
                SekkiImage[i]=SozaiView()
            }
        }

        // タッチイベントを取得
        guard let touchEvent = touches.first, let sozaiview = touchEvent.view as? SozaiView else{
            return
        }
        // ドラッグ前の座標, Swift 1.2 から
        let preStoneX = touchEvent.previousLocation(in: self.view).x
        let preStoneY = touchEvent.previousLocation(in: self.view).y
        // ドラッグ後の座標
        let newStoneX = touchEvent.location(in: self.view).x
        let newStoneY = touchEvent.location(in: self.view).y
            // ドラッグしたx座標の移動距離
        let Stonedx = newStoneX - preStoneX
        // ドラッグしたy座標の移動距離
        let Stonedy = newStoneY - preStoneY
        // 移動分を反映させる
        sozaiview.frame.origin.x += Stonedx
        sozaiview.frame.origin.y += Stonedy
        
    }

５　https://github.com/ShunHigu/OriginalNew

**Comment by yukiny0811** [comment, marked as answer]
```.swift
class TouchImageView: UIImageView {
    override init(image: UIImage?) {
        super.init(image: image)
        self.isUserInteractionEnabled = true
    }
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesMoved(touches, with: event)
        next?.touchesMoved(touches, with: event)
        let touch = touches.first!
        let prevPos = touch.previousLocation(in: self.superview)
        let nowPos = touch.location(in: self.superview)
        let diff = nowPos! - prevPos!
        self.layer.position += diff
    }
}

extension CGPoint: AdditiveArithmetic {
    public static func - (lhs: CGPoint, rhs: CGPoint) -> CGPoint {
        return CGPoint(x: lhs.x - rhs.x, y: lhs.y - rhs.y)
    }
    public static func + (lhs: CGPoint, rhs: CGPoint) -> CGPoint {
        return CGPoint(x: lhs.x + rhs.x, y: lhs.y + rhs.y)
    }
    func toCGSize() -> CGSize {
        return CGSize(width: self.x, height: self.y)
    }
}
```

**Reply by yukiny0811** [reply]
zoomで解決
